desktop_version/src/Input.cpp

```diff
@@ -1897,83 +1897,106 @@ void gameinput()
                 }
-                else
+                else if (game.press_right)
+                {
+                    obj.entities[ie].ax = 3;
+                    obj.entities[ie].dir = 1;
+                }
+            }
+        }
+
+        if (game.press_left)
+        {
+            game.tapleft++;
+        }
+        else
+        {
+            if (game.tapleft <= 4 && game.tapleft > 0)
+            {
+                for (size_t ie = 0; ie < obj.entities.size(); ++ie)
                 {
-                    if (game.tapleft <= 4 && game.tapleft > 0)
+                    if (obj.entities[ie].rule == 0)
                     {
                         if (obj.entities[ie].vx < 0.0f)
                         {
                             obj.entities[ie].vx = 0.0f;
                         }
                     }
-                    game.tapleft = 0;
-                }
-                if (game.press_right)
-                {
-                    game.tapright++;
                 }
-                else
+            }
+            game.tapleft = 0;
+        }
+        if (game.press_right)
+        {
+            game.tapright++;
+        }
+        else
+        {
+            if (game.tapright <= 4 && game.tapright > 0)
+            {
+                for (size_t ie = 0; ie < obj.entities.size(); ++ie)
                 {
-                    if (game.tapright <= 4 && game.tapright > 0)
+                    if (obj.entities[ie].rule == 0)
                     {
                         if (obj.entities[ie].vx > 0.0f)
                         {
                             obj.entities[ie].vx = 0.0f;
                         }
                     }
-                    game.tapright = 0;
-                }
-
-
-                if(game.press_left)
-                {
-                    obj.entities[ie].ax = -3;
-                    obj.entities[ie].dir = 0;
-                }
-                else if (game.press_right)
-                {
-                    obj.entities[ie].ax = 3;
-                    obj.entities[ie].dir = 1;
                 }
+            }
+            game.tapright = 0;
+        }
 
-                if (!game.press_action)
-                {
-                    game.jumppressed = 0;
-                    game.jumpheld = false;
-                }
+        if (!game.press_action)
+        {
+            game.jumppressed = 0;
+            game.jumpheld = false;
+        }
 
-                if (game.press_action && !game.jumpheld)
-                {
-                    game.jumppressed = 5;
-                    game.jumpheld = true;
-                }
+        if (game.press_action && !game.jumpheld)
+        {
+            game.jumppressed = 5;
+            game.jumpheld = true;
+        }
 
-                if (game.jumppressed > 0)
+        if (game.jumppressed > 0)
+        {
+            game.jumppressed--;
+            if (obj.entities[obj.getplayer()].onground>0 && game.gravitycontrol == 0)
```

Comment:
Why the `obj.getplayer()` here? This will only check if the first player entity is on the ground, and none of the other ones.

I tested it just now, and this change breaks mid-air flip levels (such as [this one](https://distractionware.com/forum/index.php?topic=3556.0)), where a duplicate player entity is put inside a box, thus making their `onground` and `onroof` attributes almost always 2, letting the first player entity flip in mid-air. But if you only check the `onground` and `onroof` of the first player entity, then of course they will no longer be able to flip in mid-air.

Also, I'd prefer it if you did an `INBOUNDS()` check on `obj.getplayer()` to uphold the norm of defensive programming.